File: go/tracing.go
package medallion

import (
	"context"
	"net/http"
	"strings"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/trace"
)

const (
	defaultTracerName         = "github.com/jim-technologies/medallion-sdk/go"
	defaultSpanPrefix         = "medallion"
	requestSpanFailureMessage = "Medallion request failed."
)

// requestSpanFailure is deliberately detached from the caller-facing error.
// Error messages can contain server responses, credentials, or event payloads,
// none of which belong in telemetry.
type requestSpanFailure struct{}

func (requestSpanFailure) Error() string {
	return requestSpanFailureMessage
}

func (c TracingConfig) enabled() bool {
	return c.Enabled || c.Tracer != nil
}

func (c TracingConfig) tracer() trace.Tracer {
	if c.Tracer != nil {
		return c.Tracer
	}
	name := c.TracerName
	if name == "" {
		name = defaultTracerName
	}
	return otel.Tracer(name)
}

func (c TracingConfig) spanPrefix() string {
	if c.SpanPrefix != "" {
		return c.SpanPrefix
	}
	return defaultSpanPrefix
}

func (c TracingConfig) startRequestSpan(ctx context.Context, method, path string) (context.Context, trace.Span) {
	if !c.enabled() {
		return ctx, nil
	}
	return c.tracer().Start(
		ctx,
		c.spanPrefix()+" "+method+" "+path,
		trace.WithSpanKind(trace.SpanKindClient),
		trace.WithAttributes(
			attribute.String("medallion.sdk.language", "go"),
			attribute.String("medallion.request.path", path),
			attribute.String("http.request.method", method),
		),
	)
}

func (c TracingConfig) inject(ctx context.Context, header http.Header) {
	if !c.enabled() {
		return
	}
	// Propagate only standard tracing headers from the process-wide propagator.
	// This request starts with no caller headers, so an allowlist keeps custom
	// propagators from smuggling authentication, scope, or transport metadata
	// into an SDK-owned request.
	injected := make(http.Header)
	otel.GetTextMapPropagator().Inject(ctx, propagation.HeaderCarrier(injected))
	for name, values := range injected {
		if !allowedTracingHeader(name) {
			continue
		}
		header[name] = append([]string(nil), values...)
	}
}

func allowedTracingHeader(name string) bool {
	name = strings.ToLower(name)
	switch name {
	case "b3", "baggage", "grpc-trace-bin", "traceparent", "tracestate", "uber-trace-id", "x-amzn-trace-id", "x-cloud-trace-context":
		return true
	default:
		return strings.HasPrefix(name, "uberctx-") || strings.HasPrefix(name, "x-b3-")
	}
}

func setRequestSpanResponse(span trace.Span, status int, requestID string) {
	if span == nil {
		return
	}
	span.SetAttributes(attribute.Int("http.response.status_code", status))
	if requestID != "" {
		span.SetAttributes(attribute.String("medallion.request_id", requestID))
	}
	if status < 200 || status >= 300 {
		span.SetStatus(codes.Error, "")
	}
}

func recordRequestSpanError(span trace.Span, err error) {
	if span == nil || err == nil {
		return
	}
	span.RecordError(requestSpanFailure{})
	span.SetStatus(codes.Error, "")
}

func setRequestSpanOK(span trace.Span) {
	if span != nil {
		span.SetStatus(codes.Ok, "")
	}
}

File: src/tracing.ts
import {
  context,
  propagation,
  type Span,
  SpanKind,
  SpanStatusCode,
  type Tracer,
  trace,
} from "@opentelemetry/api";

export interface TracingOptions {
  enabled?: boolean;
  tracer?: Tracer;
  tracerName?: string;
  spanPrefix?: string;
}

export type TracingConfig = boolean | TracingOptions;

export interface NormalizedTracing {
  enabled: boolean;
  tracer: Tracer;
  spanPrefix: string;
}

export interface TraceRequestOptions<TResponse> {
  method: string;
  path: string;
  run: (span?: Span) => Promise<TResponse>;
  headers: Headers;
}

const DEFAULT_TRACER_NAME = "@jimtech/medallion";
const DEFAULT_SPAN_PREFIX = "medallion";

export function normalizeTracing(config?: TracingConfig): NormalizedTracing {
  if (config === undefined || config === false) {
    return {
      enabled: false,
      tracer: trace.getTracer(DEFAULT_TRACER_NAME),
      spanPrefix: DEFAULT_SPAN_PREFIX,
    };
  }

  if (config === true) {
    return {
      enabled: true,
      tracer: trace.getTracer(DEFAULT_TRACER_NAME),
      spanPrefix: DEFAULT_SPAN_PREFIX,
    };
  }

  const enabled = config.enabled ?? config.tracer !== undefined;
  return {
    enabled,
    tracer:
      config.tracer ??
      trace.getTracer(config.tracerName ?? DEFAULT_TRACER_NAME),
    spanPrefix: config.spanPrefix ?? DEFAULT_SPAN_PREFIX,
  };
}

export async function traceRequest<TResponse>(
  tracing: NormalizedTracing,
  options: TraceRequestOptions<TResponse>,
): Promise<TResponse> {
  if (!tracing.enabled) {
    return options.run(undefined);
  }

  const span = tracing.tracer.startSpan(
    `${tracing.spanPrefix} ${options.method} ${options.path}`,
    {
      kind: SpanKind.CLIENT,
      attributes: {
        "medallion.sdk.language": "typescript",
        "medallion.request.path": options.path,
        "http.request.method": options.method,
      },
    },
  );

  return context.with(trace.setSpan(context.active(), span), async () => {
    propagation.inject(context.active(), options.headers, {
      set: (carrier, key, value) => {
        carrier.set(key, value);
      },
    });

    try {
      const response = await options.run(span);
      span.setStatus({ code: SpanStatusCode.OK });
      return response;
    } catch (error) {
      recordSpanException(span, error);
      throw error;
    } finally {
      span.end();
    }
  });
}

export function setResponseSpanAttributes(
  span: Span | undefined,
  status: number,
  requestId?: string,
): void {
  if (span === undefined) {
    return;
  }

  span.setAttribute("http.response.status_code", status);
  if (requestId !== undefined) {
    span.setAttribute("medallion.request_id", requestId);
  }
}

function recordSpanException(span: Span, error: unknown): void {
  span.recordException(error instanceof Error ? error : String(error));
  span.setStatus({
    code: SpanStatusCode.ERROR,
    message: error instanceof Error ? error.message : undefined,
  });
}
